File: python/77_Combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def combine(start = 1, curr = []):
            if len(curr) == k:
                res.append(curr)
            else:            
                for i in range(start, n + 1):
                    combine(i + 1, curr + [i])
            return res
        
        res = []
        return combine()

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = []
        def helper(current_list):
            if len(current_list) == k:
                result.append(current_list)
            else:
                initial = current_list[-1]+1 if len(current_list) > 0 else 1
                for i in range(initial, n+1):
                    helper(current_list+[i])
            return result
        return helper([])

File: swift/77_Combinations.swift
class Solution {
    func combine(_ n: Int, _ k: Int) -> [[Int]] {
        func helper(_ start: Int, _ currentCombination: [Int]) {
            if currentCombination.count == k {
                combinations.append(currentCombination)
            } else if start <= n {
                for i in start...n {
                    helper(i + 1, currentCombination + [i])
                }
            }
        }
        
        var combinations = [[Int]]()
        helper(1, [Int]())
        return combinations
    }
}
